### ISSUE TITLE
Multipart upload to AWS S3 does not seem to work

### ISSUE THREAD

**Issue by okia** [minio/minio-go#1025]
# Repro steps:
## Prerequisites:
https://github.com/minio/minio-go , version 6.0.6

## Repro steps for FPutObj
This repro scenario uses FPutObj function in order to upload BIG file to AWS S3.
- Step 1: Prepare _big enough_ file to be uploaded to AWS, say 200MB+ 
- Step 2: Start with [fputencrypted-object.go](https://github.com/minio/minio-go/blob/master/examples/s3/fputencrypted-object.go)
- Step 3: Replace [line 50](https://github.com/minio/minio-go/blob/master/examples/s3/fputencrypted-object.go#L50) with 
the following one: `encryption := encrypt.NewSSE()` . This is because we want server-side encryption. You'll also have to comment out [line 47](https://github.com/minio/minio-go/blob/master/examples/s3/fputencrypted-object.go#L47) in order to compile.
- Step 4: Replace placeholders in the [fputencrypted-object.go](https://github.com/minio/minio-go/blob/master/examples/s3/fputencrypted-object.go) [for AWS keys and/or secrets, bucket, object names, files etc.] with your actual values. 
- Note: instead of ["my-testfile"](https://github.com/minio/minio-go/blob/master/examples/s3/fputencrypted-object.go#L44) use path to the file prepared on Step 1.
- Step 5: Compile and execute example.
#### Expected: 
file successfully uploaded into AWS S3 bucket and server-side encrypted (using AES-256)
#### Actual result: 
- program produces the following error: `x-amz-server-side-encryption header is not supported for this operation`
- file is not uploaded

## Repro steps for PutObj
This repro scenario uses PutObj function in order to upload file to AWS S3.

- Step 1: Start with [put-encrypted-object.go](https://github.com/minio/minio-go/blob/master/examples/s3/put-encrypted-object.go)
- Step 2: Replace [line 64](https://github.com/minio/minio-go/blob/master/examples/s3/put-encrypted-object.go#L64) with 
 the following one: `encryption := encrypt.NewSSE()`. This is because we want server-side encryption. You'll also have to comment out [line 61](https://github.com/minio/minio-go/blob/master/examples/s3/put-encrypted-object.go#L61) in order to compile.
- Step 3: Replace 'fstat.Size()' in [line 67](https://github.com/minio/minio-go/blob/master/examples/s3/put-encrypted-object.go#L67) with '-1' to force multipart uploading.
- Step 4: Replace placeholders in the [put-encrypted-object.go](https://github.com/minio/minio-go/blob/master/examples/s3/put-encrypted-object.go) [for AWS keys and/or secrets, bucket, object names, files etc.] with your actual values. 
- Step 5: Compile and execute example.
#### Expected: 
file successfully uploaded into AWS S3 bucket and server-side encrypted (using AES-256)
#### Actual result: 
- program produces the following error: `x-amz-server-side-encryption header is not supported for this operation`
- file is not uploaded
### Note:
Interestingly enough, if one skips Step 3 and uses for uploading _small enough_ file, the program succeeds.
In other words, only multipart appears to be broken.

**Comment by kannappanr** [comment]
@okia Thanks for filing this issue. `minio-go` is probably sending `x-amz-server-side-encryption` headers during each multi-part `PUT` call. AWS used to ignore those headers. They probably are throwing an error now.

**Comment by okia** [comment]
Hello!
Just curious if you had any chance to test the fix?
It is still does not work for, even when I am using new version comprising the fix :(
Sincerely,
okia

**Comment by harshavardhana** [comment]
> Hello!
> Just curious if you had any chance to test the fix?
> It is still does not work for, even when I am using new version comprising the fix :(
> Sincerely,
> okia

yes it works fine @okia you need to use latest master not 6.0.6

**Comment by kannappanr** [comment]
Yes @okia I verified the fix with the following code.

```
package main

import (
	"bytes"
	"io/ioutil"
	"log"
	"os"

	"github.com/minio/minio-go/pkg/encrypt"

	minio "github.com/minio/minio-go"
)

func main() {
	accessKey := "accessKey"
	secretKey := "secretKey"
	bucketName := "yourbucket"
	objectName := "yourobject"
	path := "path/to/your/file"

	minioClient, err := minio.New("s3.amazonaws.com", accessKey, secretKey, true)
	if err != nil {
		log.Fatalln(err)
	}

	encryption := encrypt.NewSSE()
	minioClient.TraceOn(os.Stderr)
	_, err = minioClient.FPutObject(bucketName, objectName, path, minio.PutObjectOptions{
		ServerSideEncryption: encryption,
	})
	if err != nil {
		log.Fatalln(err)
	}

	reader, err := minioClient.GetObject(bucketName, objectName, minio.GetObjectOptions{ServerSideEncryption: encryption})
	if err != nil {
		log.Fatalln(err)
	}
	defer reader.Close()

	decBytes, err := ioutil.ReadAll(reader)
	if err != nil {
		log.Fatalln(err)
	}
	f, err := os.Open(path)
	origBytes, err := ioutil.ReadAll(f)
	if err != nil {
		log.Fatalln(err)
	}
	if !bytes.Equal(decBytes, origBytes) {
		log.Fatalln("error in matching")
	}
}

```

The latest master has the fix.

**Comment by okia** [comment]
Thank you so much @harshavardhana for you promptly answer!
Interesting point!
I git-pulled `//github.com/minio/minio-go` before recompiling my examples, is it enough to guarantee the latest minio version? Sorry, I am relatively new to golang ...
Sincerely,
okia

**Comment by kannappanr** [comment]
One way is to make sure that it is in your `gopath`. For example, my `gopath` is set to `/home/user/gopath` and the source is in `/home/user/gopath/src/github.com/minio/minio-go`

**Comment by okia** [comment]
@kannappanr 
Thanks a lot!
My particular $GOPATH location is different from yours, but essential part is all the same: I have my minio under $GOPATH/src (in `$(GOPATH)/src/github.com/minio/minio-go`) and I git-pulled this folder to pick up the fix. And yet ... it does not seem to work for me.
Let me try your [code snippet](https://github.com/minio/minio-go/issues/1025#issuecomment-423004031) and see how it works.

**Comment by harshavardhana** [comment]
> @kannappanr
> Thanks a lot!
> My particular $GOPATH location is different from yours, but essential part is all the same: I have my minio under $GOPATH/src (in `$(GOPATH)/src/github.com/minio/minio-go`) and I git-pulled this folder to pick up the fix. And yet ... it does not seem to work for me.
> Let me try your [code snippet](https://github.com/minio/minio-go/issues/1025#issuecomment-423004031) and see how it works.

@okia try doing `go get -u github.com/minio/minio-go` and then build your examples. Also paste `git log | head -1` output.
